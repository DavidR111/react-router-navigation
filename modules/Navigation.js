/* @flow */
/* eslint no-duplicate-imports: 0 */
/* eslint react/no-children-prop: 0 */

import React, { Component, createElement } from 'react'
import type { NavigationSceneRendererProps } from 'react-navigation/src/TypeDefinition'
import type { NavBarProps, CardRendererProps } from './TypeDefinitions'
import CardStack from './CardStack'
import DefaultRenderer from './DefaultRenderer'
import NavBar from './NavBar'
import StackUtils from './StackUtils'

type SceneRendererProps = CardRendererProps & NavigationSceneRendererProps

type Props = NavBarProps & {
  children: Array<React$Element<any>>,
  render: (props: SceneRendererProps) => React$Element<any>,
}

class Navigation extends Component<void, Props, void> {

  props: Props

  renderHeader = (props: SceneRendererProps): ?React$Element<any> => {
    // Get current card $FlowFixMe
    const { cards, scene: { route } } = props
    const card = StackUtils.get(cards, route)
    if (!card) return null
    // Get nav bar props
    const navBarProps = { ...this.props, ...card }
    // Hide nav bar
    if (navBarProps.hideNavBar) return null
    // Render custom nav bar
    if (navBarProps.renderNavBar) {
      return navBarProps.renderNavBar(navBarProps)
    }
    // Else return default <NavBar /> component
    return <NavBar {...navBarProps} />
  }

  renderScene = (props: SceneRendererProps): ?React$Element<any> => {
    // Get card $FlowFixMe
    const { cards, scene: { route } } = props
    const card = StackUtils.get(cards, route)
    if (!card) return null
    // Render view
    if (card.render) return card.render(props)
    else if (card.children) return card.children(props)
    else if (card.component) return createElement(card.component, props)
    return null
  }

  render(): React$Element<any> {
    const { children, ...props } = this.props
    return (
      <CardStack
        {...props}
        children={children}
        render={(ownProps) => (
          <DefaultRenderer
            {...props}
            {...ownProps}
            renderScene={this.renderScene}
            renderHeader={this.renderHeader}
          />
        )}
      />
    )
  }

}

export default Navigation
