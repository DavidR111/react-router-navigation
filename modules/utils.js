/* @flow */
/* eslint no-param-reassign: 0 */

import { Children } from 'react'
import { matchPattern } from 'react-router'
import type { NavigationRoute } from 'react-native/Libraries/NavigationExperimental/NavigationTypeDefinition'
import type { Card, MatchCardProps, Tab, MatchTabProps } from './StackTypeDefinitions'

/**
 * Extract current match from all matches
 * and return a route with key prop
 */
export function getCurrentRoute(
  matchs: Array<{ pattern: string, exactly?: boolean } & Object>,
  parent: ?string,
  location: { pathname: string },
): ?NavigationRoute {
  const route = matchs
    .find(({ pattern, exactly }) => {
      return matchPattern(pattern, location, exactly || false, parent)
    })
  if (!route) return null
  return { key: route.pattern }
}


/**
 * Build an item stack by extracting
 * all valid props
 */
export function buildItemStack<Item>(
  children: Array<React$Element<Item>>,
  advancedProps: Array<string>,
  matchComponent: string,
): Array<Item & { key: string }> {
  // Define required props
  const defaultProps = ['component', 'render', 'pattern', 'exactly']
  // Map all children
  return Children.toArray(children).reduce((items, child) => {
    // child === <Match /> || child === <MatchCard />
    const displayName = child.displayName || child.type.displayName || child.type.name
    const isAdvancedMatch = matchComponent === displayName
    if (displayName !== 'Match' && !isAdvancedMatch) return items
    // Build item
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'pattern') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (isAdvancedMatch && (key === 'render' || key === 'component')) {
        props[key] = () => child
      } else if (
        defaultProps.includes(key) ||
        (isAdvancedMatch && advancedProps.includes(key))
      ) {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.key) return items
    return [...items, item]
  }, [])
}


/**
 * Build a card stack
 */
export function buildCards(children: Array<React$Element<MatchCardProps>>): Array<Card> {
  return buildItemStack(children, ['title', 'hideNavBar'], 'MatchCard')
}


/**
 * Build a tab stack
 */
export function buildTabs(children: Array<React$Element<MatchTabProps>>): Array<Tab> {
  return buildItemStack(children, ['label', 'tabBarStyle'], 'MatchTab')
}
