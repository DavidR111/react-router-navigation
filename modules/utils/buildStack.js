/* @flow */
/* eslint no-param-reassign: 0*/

import { Children, cloneElement } from 'react'

export default function buildStack<Item>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        props.key = child.props[key]
        props[key] = child.props[key]
      } else if (key === 'render' || key === 'component' || key === 'children') {
        const renderMethod = (ownProps: any) => cloneElement(child, ownProps)
        Object.defineProperty(renderMethod, 'name', { value: key })
        props[key] = renderMethod
      } else {
        props[key] = child.props[key]
      }
      return props
    }, {})
    if (!item.path) return stack
    return [...stack, item]
  }, [])
}
