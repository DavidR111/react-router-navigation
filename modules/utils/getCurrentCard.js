/* @flow */

import type { Route, Cards, Card } from './../TypeDefinitions'
import normalizeRoute from './normalizeRoute'

export default function getCurrentCard(cards: Cards, route: Route): ?Card {
  return cards.find((card) => {
    return normalizeRoute(route).key === card.key
  })
}
