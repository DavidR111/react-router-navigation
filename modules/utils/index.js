/* @flow */

export { default as buildStack } from './buildStack'
export { default as getCurrentCard } from './getCurrentCard'
export { default as getCurrentRoute } from './getCurrentRoute'
export { default as normalizeRoute } from './normalizeRoute'
export { default as shouldStackUpdate } from './shouldStackUpdate'
