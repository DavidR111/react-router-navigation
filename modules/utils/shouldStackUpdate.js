/* @flow */
/* eslint no-duplicate-imports: 0*/

import { matchPath } from 'react-router'
import type { RouterHistory, Location } from 'react-router'
import type { Card } from './../TypeDefinitions'

export default function (
  currentCard: Card,
  nextCard: Card,
  routerHistory: RouterHistory,
  previousLocation: Location,
): boolean {
  const { entries, location, index } = routerHistory
  if (entries === undefined || index === undefined) return false
  const previousEntry = entries[index - 1]
  const currentEntry = entries[index]
  const nextEntry = entries[index + 1]
  const matchCurrentRoute = matchPath(location.pathname, currentCard.path, currentCard)
  const matchNextRoute = matchPath(location.pathname, nextCard.path, nextCard)
  return (
    // URL changes
    (previousLocation.pathname !== location.pathname) &&
    // case 1) basic pathname
    ((currentCard.key !== nextCard.key) ||
    // case 2) Pathname with query params
    // Ex: with same path article/:id,
    //     pathname article/2 !== article/3
    (matchCurrentRoute !== null && matchNextRoute !== null &&
     Object.keys(matchCurrentRoute.params).length !== 0 &&
     Object.keys(matchNextRoute.params).length !== 0 &&
     ((previousEntry && currentEntry.pathname !== previousEntry.pathname) ||
      (nextEntry && currentEntry.pathname !== nextEntry.pathname))
    ))
  )
}
