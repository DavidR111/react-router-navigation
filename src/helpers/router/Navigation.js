/* @flow */

import React, { Component, createElement } from 'react'
import { NavigationExperimental, BackAndroid, Platform, View } from 'react-native'
import NavBar from '@components/NavBar'
import CardStack from './CardStack'

const {
  CardStack: NavigationCardStack,
} = NavigationExperimental

class Navigation extends Component {

  componentDidMount() {
    BackAndroid.addEventListener('hardwareBackPress', () => {
      this.back()
      return true
    })
  }

  componentWillUnmount() {
    BackAndroid.removeEventListener('hardwareBackPress')
  }

  back = (): void => {
    this.props.onNavigate({ type: 'pop' })
  }

  renderNavBar = (sceneProps: Object): ReactElement<any> | null => {
    if (Platform.OS == 'ios') {
      return (
        <NavBar
          onNavigate={this.props.onNavigate}
          {...sceneProps}
        />
      )
    }
    return null
  }

  renderScene = (sceneProps: Object): ReactElement<any> => {
    const scene = sceneProps.scene
    const route = scene.route
    return (
      <View style={{ flex: 1 }}>
        {Platform.OS == 'android' &&
          <NavBar
            onNavigate={(action) => this.props.onNavigate(action)}
            {...sceneProps}
          />
        }
        {route.component
           ? createElement(route.component)
           : null
        }
      </View>
    )
  }

  render() {
    const { appNavigationState } = this.props
    const scenes = appNavigationState
    return (
      <View style={{ flex: 1 }}>
        <CardStack
          direction={Platform.OS == 'android' ? 'vertical' : 'horizontal'}
          navigationState={scenes}
          onNavigateBack={this.back}
          renderOverlay={(sceneProps) => this.renderNavBar(sceneProps)}
          renderScene={(sceneProps) => this.renderScene(sceneProps)}
        />
      </View>
    )
  }

}

export default Navigation
