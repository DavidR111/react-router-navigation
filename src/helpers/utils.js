/* @flow */

import _ from 'lodash'
import type { NavigationState, NavigationRoute, NavigationScene } from './../types'

export function extractScenes(
  children: Array<React$Element<NavigationScene>>
): Array<NavigationRoute> {
  return children.map((child) => ({
    ...child.props,
    key: child.key,
  }))
}

export function normalizePath(path: string): string {
  return path
    .split('.')
    .filter((_path) => _path)
    .map((_path) => `routes[${_path}]`)
    .join('.')
}

export function getSiblingScenes(state: NavigationState) {
  const normalizedPath = normalizePath(state.path.slice(0, -2))
  const path = normalizedPath
    ? `${normalizedPath}.children`
    : 'children'
  const sceneChildren = _.get(state, path)
  return extractScenes(sceneChildren)
}

export function getCurrentRoute(state: NavigationState): NavigationRoute {
  const normalizedPath = normalizePath(state.path)
  return _.get(state, normalizedPath)
}

export function findPathOfClosestTabs(state: NavigationState): string {
  const paths = state.path.split('.')
  const path = paths
    .map((_path, index) => paths.slice(0, index + 1).join('.'))
    .reverse()
    .map((_path) => ({
      path: _path,
      navigationState: _.get(state, normalizePath(_path)),
    }))
    .filter(({ navigationState }) => navigationState.tabs)
    .map((tabState) => tabState.path)
    .join('')
  return path
}
