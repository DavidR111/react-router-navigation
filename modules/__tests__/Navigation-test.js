import './__mocks__'
import React from 'react'
import { View } from 'react-native'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import { componentFactory } from './utils'
import Navigation from './../Navigation'
import renderer from 'react-test-renderer'

it('<Navigation /> renders correctly', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
      </Navigation>
    </Router>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> renders correctly with initialIndex and initialEntries prop ', () => {
  const history = createHistory({
    initialIndex: 1,
    initialEntries: ['/', '/hello', '/goodbye'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
        <Route path="/goodbye" render={componentFactory('Goodbye')} />
      </Navigation>
    </Router>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "push" action is called', () => {
  const history = createHistory()
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
      </Navigation>
    </Router>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.push('/hello')
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "goBack" action is called', () => {
  const history = createHistory({
    initialIndex: 1,
    initialEntries: ['/', '/hello'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
      </Navigation>
    </Router>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.goBack()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "go" action is called', () => {
  const history = createHistory({
    initialIndex: 2,
    initialEntries: ['/', '/hello', '/goodbye'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
        <Route path="/goodbye" render={componentFactory('Goodbye')} />
      </Navigation>
    </Router>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.go(-2)
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('<Navigation /> re-renders correctly when "replace" action is called', () => {
  const history = createHistory({
    initialIndex: 1,
    initialEntries: ['/', '/hello'],
  })
  const component = renderer.create(
    <Router history={history}>
      <Navigation hideNavBar>
        <Route exact path="/" render={componentFactory('Index')} />
        <Route path="/hello" render={componentFactory('Hello')} />
        <Route path="/goodbye" render={componentFactory('Goodbye')} />
      </Navigation>
    </Router>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  history.replace('/goodbye')
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
