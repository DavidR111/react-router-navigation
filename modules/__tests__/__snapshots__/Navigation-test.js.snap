exports[`<Navigation /> <Navigation /> re-renders correctly when "go" action is called 1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card />
      <Card>
        <Text>
          Goodbye
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "go" action is called 2`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
      <Card />
      <Card />
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "goBack" action is called 1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "goBack" action is called 2`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
      <Card />
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "push" action is called 1`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "push" action is called 2`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "replace" action is called 1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> re-renders correctly when "replace" action is called 2`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Goodbye
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> renders correctly 1`] = `
<View>
  <View>
    <View>
      <Card>
        <Text>
          Index
        </Text>
      </Card>
    </View>
  </View>
</View>
`;

exports[`<Navigation /> <Navigation /> renders correctly with initialIndex and initialEntries prop  1`] = `
<View>
  <View>
    <View>
      <Card />
      <Card>
        <Text>
          Hello
        </Text>
      </Card>
      <Card />
    </View>
  </View>
</View>
`;
